Jenkins: add configuration to verify and deploy CI jobs automatically

nixpanic opened the pull request:
### Explain the changes
1. add artefacts for configuring the Jenkins deployment in OpenShift
2. add [jjb-validate](https://jenkins-noobaa.apps.ocp.ci.centos.org/job/jjb-validate/) and [jjb-deploy](https://jenkins-noobaa.apps.ocp.ci.centos.org/job/jjb-deploy/) Jenkins Jobs Builder files
3. add Dockerfile for a jjb container to verify/deploy the `.jenkins/jobs/*.yaml` jobs

### Testing Instructions:
1. manual deployment has been done, PRs trigger jobs in the [Jenkins CentOS CI](https://jenkins-noobaa.apps.ocp.ci.centos.org/)
2. a change to the `.jenkins/jobs/*.yaml` files can be tested with the new jjb-validate job. The job can be triggered by leaving a comment with the contents `/test ci/centos/jjb-validate`
3. jobs do not report a status to the PRs _yet_, check the Jenkins WebUI for results of the tests

### TODO
1. once this PR is merged, it is needed to manually run the [jjb-deploy job](https://jenkins-noobaa.apps.ocp.ci.centos.org/job/jjb-deploy/) to update the jobs from the noobaa-core repository (now running from the source of this PR)
2. "Build & Sanity" job will fail because there is currently a conflict between Podman and minikube/Docker on CentOS 8
3. in order to have the status of the CI jobs report to PRs, credentials of a user with admin permissions need to be added to Jenkins

nixpanic committed:
jenkins: add Dockerfile to run JJB

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
jenkins: add JJB validate/deploy CI jobs

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
jenkins: add OpenShift deployment details

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
@liranmauda this is now active in Jenkins already (with minor manual modifications)

liranmauda commented on .jenkins/deploy/Makefile:
```diff
@@ -0,0 +1,8 @@
+WORKDIR ?= $(CURDIR)/../
+OUTPUT ?= $(WORKDIR)/_output/
+
+test:
+	cd $(WORKDIR) && jenkins-jobs test -o $(OUTPUT) jobs
```
```suggestion
test:
	cd $(WORKDIR) && jenkins-jobs test -o $(OUTPUT) jobs
.PHONY: test
```

liranmauda commented on .jenkins/deploy/Makefile:
```diff
@@ -0,0 +1,8 @@
+test:
+	cd $(WORKDIR) && jenkins-jobs test -o $(OUTPUT) jobs
+
+deploy:
+	cd $(WORKDIR) && jenkins-jobs update --delete-old jobs
```
```suggestion
deploy:
	cd $(WORKDIR) && jenkins-jobs update --delete-old jobs
.PHONY: deploy	
```

liranmauda commented on .jenkins/deploy/README.md:
```diff
@@ -0,0 +1,46 @@
+
+The used Jenkins environment is expected to be deployed already. This is done
+by the CentOS CI team when a [request for CI resources](ci_request) is handled.
+The deploying and configuration of Jenkins is therefor not part of this
```
```suggestion
The deploying and configuration of Jenkins is therefore not part of this
```

liranmauda commented on .jenkins/deploy/checkout-repo.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+
+fail() {
+	echo "${@}" > /dev/stderr
```
```suggestion
	echo "$(date) ${@}" > /dev/stderr
```

liranmauda commented on .jenkins/deploy/jjb.sh:
```diff
@@ -0,0 +1,77 @@
+	oc get "pod/${1}" --no-headers -o=jsonpath='{.status.phase}'
+}
+
+case "${CMD}" in
+	"validate")
+		;;
+	"deploy")
+		;;
```
What are those empty cases? Is it intentional? 
Are they only to validate that the arg is correct?

nixpanic replied:
yes, only to validate the argument

nixpanic committed:
jenkins: use GitHub web-hook to trigger jobs

The 'cron' configuration for the CI jobs is not needed anymore, the
web-hook has been configured in the GitHub repository.

Also removing the 'sandbox' option, it has been deprecated and is not
used anymore.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
jenkins: disable Build & Sanity job until it is fixed

The current Build & Sanity job in Jenkins fail to install Docker after
Podman is installed. This causes the job to abort even before testing
was done.

Updates: #6199
Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
/test ci/centos/jjb-validate

nixpanic commented:
/test ci/centos/jjb-validate

nixpanic commented:
/test ci/centos/jjb-validate

nixpanic commented:
> /test ci/centos/jjb-validate

@liranmauda maybe this does not work for me, because I am not part of the NooBaa organization. Could you try leaving such a comment?

liranmauda commented:
/test ci/centos/jjb-validate

nixpanic commented:
> /test ci/centos/jjb-validate

Ok, I'm not sure why that fails... But at least PRs that get created trigger a run of the [unit job](https://jenkins-noobaa.apps.ocp.ci.centos.org/job/noobaa-core_unit/) automatically.

nixpanic commented:
/retest ci/centos/unit

nixpanic commented:
@liranmauda this PR should be complete and can be merged. I'll open an other one to address #6199 separately.

nixpanic commented:
The (fake) CI status **ci/centos/config-check** confirms the configuration inside Jenkins is working as expected.
